### PR TITLE
receivers: update record after set change

### PR DESCRIPTION
**Pull request by hachreak**
Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

File: invenio_oaiserver/query.py [hunk 1]
```diff
@@ -67,29 +70,42 @@ def get_records(**kwargs):
...
+        yield {
+            'id': result['_id'],
+            'json': result['_source'],
+            # FIXME use ES
+            'updated': RecordMetadata.query.filter_by(
+                id=result['_id']).one().updated,
+        }
+
+
+class Pagination(object):
```
Comment (jirikuncar) [inline invenio_oaiserver/query.py]: see changes in #15 and keep Pagination as local private class for now.

File: tests/test_percolator.py [hunk 2]
```diff
@@ -72,32 +75,44 @@ def _try_populate_oaisets():
...
     record3 = Record.create({'title': 'Test3', '$schema': schema})
+    indexer.index_by_id(record3.id)
+    sleep(1)
 
     assert 'e' in record3['_oai']['sets']
     assert 'i' in record3['_oai']['sets']
     assert len(record3['_oai']['sets']) == 2
 
     record4 = Record.create({'title': 'Test4', '$schema': schema})
+    indexer.index_by_id(record4.id)
```
Comment (jirikuncar) [inline tests/test_percolator.py]: `indexer.index(record4)`

File: invenio_oaiserver/receivers.py [hunk 3]
```diff
@@ -102,3 +103,39 @@ def __call__(self, record, **kwargs):
...
             }
         })
+
+
+def update_records():  # records_info):
+    """Update oai record informations."""
+    from invenio_records.models import RecordMetadata
+    #  rinfo = list(records_info)
+    #  if rinfo:
+    records = RecordMetadata.query.all()
```
Comment (jirikuncar) [inline invenio_oaiserver/receivers.py]: you can use `scan` [1] from `elasticsearch.helpers`.

--
[1] http://elasticsearch-py.readthedocs.org/en/master/helpers.html#scan

File: tests/test_percolator.py (hunk 2 above)
Comment (jirikuncar) [inline tests/test_percolator.py]: !

**Comment by kaplun** [comment]
I see now this branch where records are touched everytime they enter or exit a given set (as it was done in Invenio-1.x). Is this what we want?

What about simply defining sets as regular queries resolved run time? Additionally @Dziolas was working on dynamic sets (i.e. sets defined at query time). @Dziolas what is that state of that?

**Comment by hachreak** [comment]
Actually, `Sets` are defined in the same way as `Collections`.
It means that they already have the `search_pattern` field for dynamic sets.
But, also as `Collections`, the set lists is saved inside the record.

**Comment by jirikuncar** [comment]
> Is this what we want?

Yes

**Comment by kaplun** [comment]
@hachreak note that by dynamic set I meant sets that are not configured but are simply defined by the client:

E.g.:

http://.../oai2d?verb=GetRecords&set=author:foo%20and%20experiment:bar

See also:
http://oai.base-search.net/index.html#dynamic-sets

This would somehow conflict with storing the name of sets directly in the record. (of course I guess both techniques could be supported with a bit of wrapping).

**Comment by jirikuncar** [comment]
IMHO Dynamic queries are equivalent to a search with different output format (e.g. MARCXML). We were implementing this package strictly according to https://www.openarchives.org/OAI/openarchivesprotocol.html. 

http://oai.base-search.net/index.html#dynamic-sets provides non-standard extension that can be implemented as custom error handler (no set found).

**Comment by kaplun** [comment]
> http://oai.base-search.net/index.html#dynamic-sets provides non-standard extension that can be implemented as custom error handler (no set found).

:+1: nice!

**Comment by hachreak** [comment]
ping @lnielsen

File: invenio_oaiserver/response.py [hunk 4]
```diff
@@ -281,7 +281,7 @@ def getrecord(**kwargs):
         e_getrecord,
         identifier=str(pid.object_uuid),
         datestamp=record.updated,
-        sets=record.json.get('_oai', []).get('sets', []),
+        sets=record.json.get('_oai', {}).get('sets', {}),
```
Comment (lnielsen) [inline invenio_oaiserver/response.py]: ``sets=record.json.get('_oai', {}).get('sets', []),``

File: invenio_oaiserver/models.py [hunk 5]
```diff
@@ -76,5 +80,16 @@ class OAISet(db.Model, Timestamp):
...
+    @hybrid_property
+    def spec(self):
+        """Get spec."""
+        return self._spec
+
+    @spec.setter
+    def spec(self, value):
+        """Set spec."""
+        if self._spec:
+            raise OAISetNotSupported("update spec not supported")
```
Comment (lnielsen) [inline invenio_oaiserver/models.py]: ``Updating spec is not supported.``

File: invenio_oaiserver/tasks.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+from flask_celeryext import RequestContextTask
+from invenio_db import db
+from invenio_records.api import Record
+from invenio_records.models import RecordMetadata
+
+from .query import get_affected_records
+
+
+@shared_task(base=RequestContextTask)
+def update_records_after_change_oaiset(**kwargs):
```
Comment (lnielsen) [inline invenio_oaiserver/tasks.py]: Explicit about which kwargs we take.

File: invenio_oaiserver/tasks.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    """Update records after a oaiset is updated, inserted or removed."""
+    records_info = list(get_affected_records(**kwargs))
```
Comment (lnielsen) [inline invenio_oaiserver/tasks.py]:
```python
for record in get_affected_records(**kwargs):
    record = Record.get(record['id])
    record.commit()
    db.session.commit()
```

File: invenio_oaiserver/query.py [hunk 8]
```diff
@@ -45,6 +46,28 @@ def match(self, record):
...
+    query = " OR ".join(filter(None, [sets, search_pattern]))
+    body = {'query': Query(query).query.accept(ElasticSearchDSL())}
+
+    response = current_search_client.search(
+        index=current_app.config['OAISERVER_RECORD_INDEX'],
+        body=body,
+    )
+
+    for result in response['hits']['hits']:
+        yield {
```
Comment (lnielsen) [inline invenio_oaiserver/query.py]: yield id instead

**Comment by hachreak** [comment]
ping @lnielsen  all green, finally! XD

File: invenio_oaiserver/errors.py [hunk 9]
```diff
@@ -29,3 +29,7 @@
 
 class OAIBadMetadataFormatError(Exception):
     """Metadata format required doesn't exist."""
+
+
+class OAISetNotSupported(Exception):
+    """Raised if is invoked a not supported function."""
```
Comment (jirikuncar) [inline invenio_oaiserver/errors.py]: What about `... it is invoked by not supported function.` ?

File: invenio_oaiserver/errors.py [hunk 10]
```diff
@@ -29,3 +29,7 @@
 
 class OAIBadMetadataFormatError(Exception):
     """Metadata format required doesn't exist."""
+
+
+class OAISetNotSupported(Exception):
```
Comment (jirikuncar) [inline invenio_oaiserver/errors.py]: Can we make the name a bit more clear with `OAISet(Update|Modification)Error`?

(exception should end with `Error`)

File: invenio_oaiserver/query.py [hunk 11]
```diff
@@ -45,6 +46,23 @@ def match(self, record):
...
 
+def get_affected_records(spec=None, search_pattern=None):
+    """Get list of affected records."""
+    sets = '_oai.sets:{0}'.format(spec) if spec else None
+    search_pattern = search_pattern if search_pattern else None
+
+    query = " OR ".join(filter(None, [sets, search_pattern]))
+    body = {'query': Query(query).query.accept(ElasticSearchDSL())}
+
+    response = current_search_client.search(
```
Comment (jirikuncar) [inline invenio_oaiserver/query.py]: please consider using `scan` helper (https://elasticsearch-py.readthedocs.org/en/master/helpers.html#elasticsearch.helpers.scan)

File: tests/conftest.py [hunk 12]
```diff
@@ -95,12 +102,14 @@ def app(request):
     with app.app_context():
         db.create_all()
         list(search.create(ignore=[400]))
-        sleep(5)
+        sleep(10)
+        current_oaiserver.unregister_signals_oaiset()
```
Comment (jirikuncar) [inline tests/conftest.py]: Do we need to call it always? The signal registration should be guarded by config option.

File: tests/conftest.py [hunk 13]
```diff
@@ -95,12 +102,14 @@ def app(request):
...
 
     def teardown():
         with app.app_context():
             list(search.delete(ignore=[404]))
             db.drop_all()
+            current_oaiserver.unregister_signals()
```
Comment (jirikuncar) [inline tests/conftest.py]: IMHO it is not needed at all during the teardown.

File: tests/test_invenio_oaiserver.py [hunk 14]
```diff
@@ -37,6 +37,13 @@ def test_version():
     assert __version__
 
 
+def test_view(app):
```
Comment (jirikuncar) [inline tests/test_invenio_oaiserver.py]: why did you move it up?

File: invenio_oaiserver/query.py (hunk 11 above)
Comment (hachreak) [inline invenio_oaiserver/query.py]: It comes from `elasticsearch.helpers`, are you sure that insert a direct dependency of `elasticsearch-py` here is a good idea? Can I access from `invenio-search`?

Comment (jirikuncar) [inline invenio_oaiserver/query.py]: It's fined to add the direct dependency to work with `current_search_client`.

**Comment by hachreak** [comment]
ping @jirikuncar @lnielsen

File: invenio_oaiserver/ext.py [hunk 15]
```diff
@@ -64,6 +65,18 @@ def register_signals(self):
                                                      weak=False)
         records_signals.before_record_update.connect(self.update_function,
                                                      weak=False)
+        #  import ipdb; ipdb.set_trace()  # XXX BREAKPOINT
```
Comment (jirikuncar) [inline invenio_oaiserver/ext.py]: FIXME

File: invenio_oaiserver/query.py [hunk 16]
```diff
@@ -45,6 +47,24 @@ def match(self, record):
         return self.query.accept(MatchUnit(record))
 
 
+def get_affected_records(spec=None, search_pattern=None):
+    """Get list of affected records."""
+    sets = '_oai.sets:{0}'.format(spec) if spec else None
```
Comment (jirikuncar) [inline invenio_oaiserver/query.py]: Shouldn't the double quotes be around the spec - `'_oai.sets:"{0}"'.format(spec)`?

File: invenio_oaiserver/tasks.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+
+@shared_task(base=RequestContextTask)
+def update_records_after_change_oaiset(spec=None, search_pattern=None):
+    """Update records after a oaiset is updated, inserted or removed."""
+    for record_id in get_affected_records(spec=spec,
+                                          search_pattern=search_pattern):
+        record = Record.get_record(id=record_id)
+        record.commit()
+        db.session.commit()
```
Comment (jirikuncar) [inline invenio_oaiserver/tasks.py]: IMHO we should have a transaction around and commit after all records are updated.

File: setup.py [hunk 18]
```diff
@@ -32,6 +32,7 @@
 history = open('CHANGES.rst').read()
 
 tests_require = [
+    'mock>=1.3.0',
```
Comment (jirikuncar) [inline setup.py]: please resort the dependencies

File: tests/conftest.py [hunk 19]
```diff
@@ -79,10 +84,15 @@ def app(request):
         SERVER_NAME='app',
         OAISERVER_RECORD_INDEX='_all',
         OAISERVER_METADATA_FORMATS=metadata_formats,
+        # Disable set signals because the celery tasks cannot be run
```
Comment (jirikuncar) [inline tests/conftest.py]: Please try to wrap the task in correct transaction.

File: invenio_oaiserver/tasks.py (hunk 17 above)
Comment (hachreak) [inline invenio_oaiserver/tasks.py]: @lnielsen can tell you more about the `commit`

File: invenio_oaiserver/models.py [hunk 20]
```diff
@@ -76,5 +80,16 @@ class OAISet(db.Model, Timestamp):
     )
     """Search pattern to get records."""
 
+    @hybrid_property
```
Comment (jirikuncar) [inline invenio_oaiserver/models.py]: simple validator would do it - see http://docs.sqlalchemy.org/en/latest/orm/mapped_attributes.html#simple-validators

File: invenio_oaiserver/receivers.py [hunk 21]
```diff
@@ -103,3 +104,26 @@ def __call__(self, record, **kwargs):
...
+def after_insert_oai_set(mapper, connection, target):
+    """Update records on Set insertion."""
+    update_records_after_change_oaiset.delay(
+        search_pattern=target.search_pattern
+    )
+
+
+def after_update_oai_set(mapper, connection, target):
+    """Update records on Set updated."""
+    kwargs = {
```
Comment (jirikuncar) [inline invenio_oaiserver/receivers.py]:
```python
update_records_after_change_oaiset.delay(
    spec=target.spec, search_pattern=target.search_pattern
)

File: setup.py [hunk 22]
```diff
@@ -37,6 +37,7 @@
     'invenio-marc21>=1.0.0a1',
     'isort>=4.2.2',
     'mock>=1.3.0',
+    'mock>=1.3.0',
```
Comment (jirikuncar) [inline setup.py]: see line 39!

File: tests/conftest.py [hunk 23]
```diff
@@ -71,6 +72,10 @@ def app(request):
 
     app = Flask('testapp', instance_path=instance_path)
     app.config.update(
+        CELERY_ALWAYS_EAGER=True,
```
Comment (jirikuncar) [inline tests/conftest.py]: please be consistent with single quotes

File: tests/conftest.py [hunk 24]
```diff
@@ -71,6 +72,10 @@ def app(request):
...
+        CELERY_CACHE_BACKEND="memory",
```
Comment (jirikuncar) [inline tests/conftest.py]: please be consistent with single quotes

File: tests/conftest.py [hunk 25]
```diff
@@ -95,7 +105,7 @@ def app(request):
     with app.app_context():
         db.create_all()
         list(search.create(ignore=[400]))
-        sleep(5)
+        sleep(10)
```
Comment (jirikuncar) [inline tests/conftest.py]: do we really need to slow down all tests? Can we create special fixture that creates Elastic indices only for tests that really need them.

File: invenio_oaiserver/tasks.py (hunk 17 above)
Comment (jirikuncar) [inline invenio_oaiserver/tasks.py]: have you tried something like:

```
affected_records = get_affected_records(spec=spec, search_pattern=search_pattern)
with db.session.begin_nested():
    for record_ids in grouper(afftected_records, 100):
        for record in RecordMetadata.filter(RecordMetadata.id.in_(record_ids)).all():
            current_oaiserver.update_function(record.json)
db.session.commit()
```

--
cc @lnielsen

**Comment by hachreak** [comment]
ping @jirikuncar @lnielsen 
celery task splitted!

File: invenio_oaiserver/query.py [hunk 26]
```diff
@@ -45,6 +47,24 @@ def match(self, record):
         return self.query.accept(MatchUnit(record))
 
 
+def get_affected_records(spec=None, search_pattern=None):
+    """Get list of affected records."""
+    sets = '_oai.sets:"{0}"'.format(spec) if spec else None
+    search_pattern = search_pattern if search_pattern else None
```
Comment (jirikuncar) [inline invenio_oaiserver/query.py]: If `search_pattern` is `None` then the OAI set is manually managed, hence it should not be updated.

Comment (hachreak) [inline invenio_oaiserver/query.py]: Please, could you fix it in your PR? https://github.com/hachreak/invenio-oaiserver/pull/1

Comment (jirikuncar) [inline invenio_oaiserver/query.py]: I will take over this PR to speed it up then.
